src/plugins/telemetry/include/plugins/telemetry/telemetry.h

```diff
@@ -363,6 +363,64 @@ class Telemetry : public PluginBase {
         uint8_t reset_counter{}; /**< @brief Estimate reset counter. */
     };
 
+    /**
+     * @brief Distance sensor information type.
+     *
+     * The updates from distance sensors.
+     */
+    struct DistanceSensor {
+        /**
+         * @brief Distance sensor types
+         */
+        enum class SensorType {
+            LASER = 0, /**< @brief Laser rangefinder, e.g. LightWare SF02/F or PulsedLight units. */
+            ULTRASOUND = 1, /**< @brief Ultrasound rangefinder, e.g. MaxBotix units. */
+            INFRARED = 2, /**< @brief Infrared rangefinder, e.g. Sharp units. */
+            RADAR = 3, /**< @brief Radar type, e.g. uLanding units. */
+            UNKNOWN = 4, /**< @brief Broken or unknown type, e.g. analog units. */
+        };
+
+        /**
+         * @brief Sensor orientation, according to its rotations
+         */
+        enum class SensorOrientation {
+            ROTATION_NONE = 0, /**< @brief Roll: 0, Pitch: 0, Yaw: 0 */
+            ROTATION_YAW_90 = 2, /**< @brief Roll: 0, Pitch: 0, Yaw: 90 */
+            ROTATION_YAW_270 = 6, /**< @brief Roll: 0, Pitch: 0, Yaw: 270 */
+            ROTATION_PITCH_180 = 12, /**< @brief Roll: 0, Pitch: 180, Yaw: 0 */
+            ROTATION_PITCH_90 = 24, /**< @brief Roll: 0, Pitch: 90, Yaw: 0 */
+            ROTATION_PITCH_270 = 25, /**< @brief Roll: 0, Pitch: 270, Yaw: 0 */
+            ROTATION_CUSTOM = 100 /**< @brief Custom orientation */
+        };
+
+        uint32_t time_boot_ms; /**< @brief  Timestamp (time since system boot). */
+        uint16_t min_distance; /**< @brief  Minimum distance the sensor can measure. */
+        uint16_t max_distance; /**< @brief  Maximum distance the sensor can measure. */
+        uint16_t current_distance; /**< @brief Current distance reading. */
+        SensorType type; /**< @brief Type of distance sensor. */
+        uint8_t id; /**< @brief Onboard ID of the sensor. */
```

Comment:
I don't like the fact that we are just 1 to 1 exposing the MAVLink message at this point. In my opinion the point of MAVSDK is to strip the information down to its essentials and present it in simplified way.

For instance here we have an `enum` for the orientation as well as a quaternion. What about translating the enum to quaternion inside the plugin and only exposing the quaternion?

Or what about using covariance as a float in meters with `NAN` for unknown instead of relying on the `uint8_t` type.

Another idea: should we return a vector of distance sensors where each ID is an entry instead of sending them one by one and having the client code to filter it?

Reply:
> Or what about using covariance as a float in meters with NAN for unknown instead of relying on the uint8_t type.

It is a good idea but I'm little confused because of in this case seems that Covariance is not the same as `Odometry` Covariance:

> Measurement variance. Max standard deviation is 6cm.

Anyway what about smth like  that?
```cpp
float covariance_m_2; /**< @brief Measurement variance in square meters. NNan if unknown. 
```
```cpp
uint8_t covariance_cm_2 = mavlink_msg_distance_sensor_get_covariance(&message);
if (covariance_cm_2 != 255) {
    distance_sensor.covariance = covariance_cm_2 * covariance_cm_2 / 1e4;
} else {
    distance_sensor.covariance = std::numeric_limits<float>::quiet_NaN
}
```

Reply:
> What about translating the enum to quaternion inside the plugin and only exposing the quaternion?

Is that does not make handling more complicated?
If users prefers to use standard orientations then handling `Quaternion` looks like a little bit overhead.

@physicsman are you have any idea for that? What way seems more useful for your purposes for example?

Reply:
> should we return a vector of distance sensors where each ID is an entry instead of sending them one by one and having the client code to filter it?

Hm. Interesting idea. But now I don't know how exactly we should handle received messages.
Return vector on each received message? If yes, so user will should check `time_boot_ms` on each vector member to check is it recently received message or not.
If we will drop old received messages so we don't need a vector anymore. 
But in C++17 we can use `std::map<uint8_t /*sensor_id*/, std::option<DistanceSensor>>`. I think that this is interesting solution.
If I understand you correctly.

Reply:
Hm. @julianoes , what are you think about some around individual subscription for each sensor?

```cpp
    std::map<uint8_t/*senor_id*/, Telemetry::distance_sensor_callback_t> _distance_sensors_subscriptions{};
```

Reply:
> What about translating the enum to quaternion inside the plugin and only exposing the quaternion?

Not sure if it helps the discussion, but I don't understand why MAVLink does what it does instead of just exposing a quaternion :thinking:. Is it legacy, like the enum was there first? Just having the quaternion would seem enough to me.

Reply:
> what are you think about some around individual subscription for each sensor?

So now I'm trying to think about how it would look like in the wrappers. We can easily auto-generate a wrapper for the version with the ID, and I believe it is easy to filter on the client side (be it in asyncio or in reactive extensions).

Now if we have a map of callbacks, that's more complicated. I guess we would need to expose a list of available sensor ids, and a way to subscribe with the id. Like `SubscribeAvailableDistanceSensors` or `FetchAvailableDistanceSensors`, and then `SubscribeDistanceSensor(id)`.

Either way, there is some logic required on the client side, because you need to select the sensor you are interested in, right?

Reply:
> Just having the quaternion would seem enough to me.

I'm not worked with Distance Sensors yet. So I worried that quaternion will overcomplicated for use.
Ok, I will implement Quaternion only now

Reply:
> Just having the quaternion would seem enough to me.

I'm not worked with Distance Sensors yet. So I worried that quaternion will overcomplicated for use.
Ok, I will implement Quaternion only now

Reply:
> Now if we have a map of callbacks, that's more complicated. 

Yes, in gRPC that's true. Ok, let it go :)

Reply:
> Ok, let it go :)

So which one do we go for? We keep the id like in MAVLink, or we go for another interface? I would be fine for both solutions, as long as we can auto-generate them. As a user, I don't think one of them is necessarily more complicated to deal with

Reply:
> So which one do we go for?

I mean that leave ID's as in MAVLink, sorry